File: packages/php/src/Seguranca/Ataques/AtaqueMensagemLonga.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;

/**
 * Testa mensagens grandes (vários blocos de 32 bytes de keystream). Erros de
 * sincronização de bloco, repetição de keystream em blocos distantes ou
 * perda de bytes aparecem só em mensagens longas - os testes de ida-e-volta
 * curtos não pegam. Também confirma que o mesmo texto com IVs diferentes
 * gera tokens diferentes.
 */
class AtaqueMensagemLonga implements AtaqueInterface
{
    public function __construct(private array $tamanhos = [1000, 10000, 100000])
    {
    }

    public function nome(): string
    {
        return 'Mensagens longas (multi-bloco)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $falhas = [];

        foreach ($this->tamanhos as $t) {
            $texto = random_bytes($t);
            try {
                $token = $alvo->encrypt($texto);
                if ($alvo->decrypt($token) !== $texto) {
                    $falhas[] = "$t bytes (conteúdo diferente)";
                }
            } catch (\Throwable $e) {
                $falhas[] = "$t bytes (erro: {$e->getMessage()})";
            }
        }

        $texto = str_repeat('A', 1000);
        $t1 = $alvo->encrypt($texto);
        $t2 = $alvo->encrypt($texto);
        if ($t1 === $t2) {
            $falhas[] = 'tokens idênticos para o mesmo texto (IV não varia)';
        }

        if (!empty($falhas)) {
            return new ResultadoAtaque(
                $this->nome(),
                vulneravel: true,
                severidade: 'critica',
                detalhes: 'Falha em: ' . implode('; ', $falhas),
            );
        }

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: false,
            severidade: 'info',
            detalhes: 'Ida-e-volta OK em ' . implode(', ', $this->tamanhos) . ' bytes; mesmo texto gera tokens distintos',
        );
    }
}

File: packages/python/src/Seguranca/Ataques/AtaqueMensagemLonga.py
"""Testa mensagens grandes (vários blocos de 32 bytes de keystream). Erros de
sincronização de bloco, repetição de keystream em blocos distantes ou
perda de bytes aparecem só em mensagens longas - os testes de ida-e-volta
curtos não pegam. Também confirma que o mesmo texto com IVs diferentes
gera tokens diferentes.

Adaptação Python: a API do alvo recebe `str`; geramos os bytes aleatórios e
os representamos via latin-1 (bijeção byte<->code point) para depois
comparar os BYTES originais com o resultado reencodado.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..Util import random_bytes


class AtaqueMensagemLonga:
    def __init__(self, tamanhos=None):
        self.tamanhos = [1000, 10000, 100000] if tamanhos is None else tamanhos

    def nome(self) -> str:
        return "Mensagens longas (multi-bloco)"

    def executar(self, alvo) -> ResultadoAtaque:
        falhas = []

        for t in self.tamanhos:
            texto_bytes = random_bytes(t)
            texto = texto_bytes.decode("latin-1")
            try:
                token = alvo.encrypt(texto)
                if alvo.decrypt(token).encode("latin-1") != texto_bytes:
                    falhas.append(f"{t} bytes (conteúdo diferente)")
            except Exception as e:
                falhas.append(f"{t} bytes (erro: {e})")

        texto = "A" * 1000
        t1 = alvo.encrypt(texto)
        t2 = alvo.encrypt(texto)
        if t1 == t2:
            falhas.append("tokens idênticos para o mesmo texto (IV não varia)")

        if falhas:
            return ResultadoAtaque(
                self.nome(),
                True,
                "critica",
                "Falha em: " + "; ".join(falhas),
            )

        return ResultadoAtaque(
            self.nome(),
            False,
            "info",
            "Ida-e-volta OK em "
            + ", ".join(str(t) for t in self.tamanhos)
            + " bytes; mesmo texto gera tokens distintos",
        )
